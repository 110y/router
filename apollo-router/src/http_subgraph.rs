use std::pin::Pin;

use apollo_router_core::prelude::*;
use async_trait::async_trait;
use derivative::Derivative;
use futures::Future;
use tower::Service;
use tracing::Instrument;
use url::Url;

/// A fetcher for subgraph data that uses http.
/// Streaming via chunking is supported.
#[derive(Derivative)]
#[derivative(Debug)]
pub struct HttpSubgraphFetcher {
    service: String,
    url: Url,
    #[derivative(Debug = "ignore")]
    http_client: reqwest_middleware::ClientWithMiddleware,
}

impl HttpSubgraphFetcher {
    /// Construct a new http subgraph fetcher that will fetch from the supplied URL.
    pub fn new(service: impl Into<String>, url: Url) -> Self {
        let service = service.into();

        HttpSubgraphFetcher {
            http_client: reqwest_middleware::ClientBuilder::new(
                reqwest::Client::builder()
                    .tcp_keepalive(Some(std::time::Duration::from_secs(5)))
                    .build()
                    .unwrap(),
            )
            .with(reqwest_tracing::TracingMiddleware)
            .with(LoggingMiddleware::new(&service))
            .build(),
            service,
            url,
        }
    }

    fn create_request(&self, request: graphql::Request) -> reqwest_middleware::RequestBuilder {
        self.http_client.post(self.url.clone()).json(&request)
    }

    async fn send_request(
        service: &str,
        req: reqwest_middleware::RequestBuilder,
    ) -> Result<bytes::Bytes, graphql::FetchError> {
        let response = req
            .send()
            .instrument(tracing::trace_span!("http-subgraph-request"))
            .await
            .map_err(|err| {
                tracing::error!(fetch_error = format!("{:?}", err).as_str());

                graphql::FetchError::SubrequestHttpError {
                    service: service.to_owned(),
                    reason: err.to_string(),
                }
            })?;

        response
            .bytes()
            .instrument(tracing::debug_span!("aggregate_response_data"))
            .await
            .map_err(|err| {
                tracing::error!(fetch_error = format!("{:?}", err).as_str());

                graphql::FetchError::SubrequestHttpError {
                    service: service.to_owned(),
                    reason: err.to_string(),
                }
            })
    }

    async fn request_stream(
        &self,
        request: graphql::Request,
    ) -> Result<bytes::Bytes, graphql::FetchError> {
        let req = self.create_request(request);
        Self::send_request(&self.service, req).await
    }

    fn map_to_graphql(
        service_name: String,
        response: bytes::Bytes,
    ) -> Result<graphql::Response, graphql::FetchError> {
        tracing::debug_span!("parse_subgraph_response").in_scope(|| {
            serde_json::from_slice::<graphql::Response>(&response).map_err(|error| {
                graphql::FetchError::SubrequestMalformedResponse {
                    service: service_name.clone(),
                    reason: error.to_string(),
                }
            })
        })
    }
}

#[async_trait]
impl graphql::Fetcher for HttpSubgraphFetcher {
    /// Using reqwest to fetch a graphql response
    async fn stream(
        &self,
        request: graphql::Request,
    ) -> Result<graphql::Response, graphql::FetchError> {
        let service_name = self.service.to_string();
        let response = self.request_stream(request).await?;
        Self::map_to_graphql(service_name, response)
    }
}

impl Service<graphql::Request> for HttpSubgraphFetcher {
    type Response = graphql::Response;

    type Error = graphql::FetchError;

    type Future = Pin<Box<dyn Future<Output = Result<Self::Response, Self::Error>>>>;

    fn poll_ready(
        &mut self,
        _cx: &mut std::task::Context<'_>,
    ) -> std::task::Poll<Result<(), Self::Error>> {
        std::task::Poll::Ready(Ok(()))
    }

    fn call(&mut self, request: graphql::Request) -> Self::Future {
        let service_name = self.service.to_string();

        // separate the request creation to avoid holding a reference to the fetcher
        // in the future returned by the service
        let req = self.create_request(request);

        Box::pin(async move {
            let response = Self::send_request(&service_name, req).await;
            match response {
                Err(e) => Err(e),
                Ok(response) => Self::map_to_graphql(service_name, response),
            }
        })
    }
}

struct LoggingMiddleware {
    service: String,
}

impl LoggingMiddleware {
    fn new(service: impl Into<String>) -> Self {
        Self {
            service: service.into(),
        }
    }
}

#[async_trait::async_trait]
impl reqwest_middleware::Middleware for LoggingMiddleware {
    async fn handle(
        &self,
        req: reqwest::Request,
        extensions: &mut task_local_extensions::Extensions,
        next: reqwest_middleware::Next<'_>,
    ) -> reqwest_middleware::Result<reqwest::Response> {
        tracing::trace!("Request to service {}: {:?}", self.service, req);
        let res = next.run(req, extensions).await;
        tracing::trace!("Response from service {}: {:?}", self.service, res);
        res
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use httpmock::Method::POST;
    use httpmock::{MockServer, Regex};
    use serde_json::json;
    use test_log::test;

    #[test(tokio::test)]
    async fn test_non_chunked() -> Result<(), Box<dyn std::error::Error>> {
        let expected_response = graphql::Response::builder()
            .data(json!({
              "allProducts": [
                {
                  "variation": {
                    "id": "OSS"
                  },
                  "id": "apollo-federation"
                },
                {
                  "variation": {
                    "id": "platform"
                  },
                  "id": "apollo-studio"
                }
              ]
            }))
            .build();

        let server = MockServer::start();
        let mock = server.mock(|when, then| {
            when.method(POST)
                .path("/graphql")
                .body_matches(Regex::new(".*").unwrap());
            then.status(200)
                .header("Content-Type", "application/json")
                .json_body_obj(&expected_response);
        });
        let fetcher =
            HttpSubgraphFetcher::new("products", Url::parse(&server.url("/graphql")).unwrap());
        let response = fetcher
            .stream(
                graphql::Request::builder()
                    .query(r#"{allProducts{variation {id}id}}"#)
                    .build(),
            )
            .await
            .unwrap();

        assert_eq!(response, expected_response);
        mock.assert();
        Ok(())
    }
}
