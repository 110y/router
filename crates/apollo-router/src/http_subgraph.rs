use apollo_router_core::prelude::*;
use bytes::Bytes;
use futures::prelude::*;
use std::pin::Pin;

type BytesStream = Pin<
    Box<dyn futures::Stream<Item = Result<bytes::Bytes, graphql::FetchError>> + std::marker::Send>,
>;

/// A fetcher for subgraph data that uses http.
/// Streaming via chunking is supported.
#[derive(Debug)]
pub struct HttpSubgraphFetcher {
    service: String,
    url: String,
    http_client: reqwest::Client,
}

impl HttpSubgraphFetcher {
    /// Construct a new http subgraph fetcher that will fetch from the supplied URL.
    pub fn new(service: String, url: String) -> Self {
        HttpSubgraphFetcher {
            service,
            url,
            http_client: reqwest::Client::builder()
                .tcp_keepalive(Some(std::time::Duration::from_secs(5)))
                .build()
                .unwrap(),
        }
    }

    fn request_stream(&self, request: graphql::Request) -> BytesStream {
        // Perform the actual request and start streaming.
        // Reqwest doesn't care if there is only one response, in this case it'll be a stream of one element.
        let service = self.service.to_owned();
        self.http_client
            .post(self.url.clone())
            .json(&request)
            .send()
            // We have a future for the response, convert it to a future of the stream.
            .map_ok(|r| r.bytes_stream().boxed())
            // Convert the entire future to a stream, at this point we have a stream of a result of a single stream
            .into_stream()
            // Flatten the stream
            .flat_map(|result| match result {
                Ok(s) => s,
                Err(err) => stream::iter(vec![Err(err)]).boxed(),
            })
            .map_err(
                move |err: reqwest::Error| graphql::FetchError::SubrequestHttpError {
                    service: service.to_owned(),
                    reason: err.to_string(),
                },
            )
            .boxed()
    }

    fn map_to_graphql(&self, bytes_stream: BytesStream) -> graphql::ResponseStream {
        // Map the stream of bytes to our response type.
        let service = self.service.to_owned();
        let response_stream = bytes_stream
            // We want to know if this is the primary response or not, so attach the response count
            .enumerate()
            .map(move |(index, result)| {
                let service = service.to_owned();
                match result {
                    Ok(bytes) => to_response(service, &bytes, index == 0),
                    Err(err) => err.to_response(index == 0),
                }
            });
        response_stream.boxed()
    }
}

fn to_response(service: impl Into<String>, bytes: &Bytes, primary: bool) -> graphql::Response {
    serde_json::from_slice::<graphql::Response>(bytes)
        .map_err(
            move |err| graphql::FetchError::SubrequestMalformedResponse {
                service: service.into(),
                reason: err.to_string(),
            },
        )
        .unwrap_or_else(|err| err.to_response(primary))
}

impl graphql::Fetcher for HttpSubgraphFetcher {
    /// Using reqwest fetch a stream of graphql results.
    fn stream(&self, request: graphql::Request) -> graphql::ResponseStream {
        let bytes_stream = self.request_stream(request);
        self.map_to_graphql(bytes_stream)
    }
}

#[cfg(test)]
mod tests {
    use httpmock::Method::POST;
    use httpmock::{MockServer, Regex};
    use serde_json::json;

    use super::*;

    #[tokio::test]
    async fn test_non_chunked() -> Result<(), Box<dyn std::error::Error>> {
        let response = graphql::Response::builder()
            .data(json!({
              "allProducts": [
                {
                  "variation": {
                    "id": "OSS"
                  },
                  "id": "apollo-federation"
                },
                {
                  "variation": {
                    "id": "platform"
                  },
                  "id": "apollo-studio"
                }
              ]
            }))
            .build();

        let server = MockServer::start();
        let mock = server.mock(|when, then| {
            when.method(POST)
                .path("/graphql")
                .body_matches(Regex::new(".*").unwrap());
            then.status(200)
                .header("Content-Type", "application/json")
                .json_body_obj(&response);
        });
        let fetcher = HttpSubgraphFetcher::new("products".into(), server.url("/graphql"));
        let collect = fetcher
            .stream(
                graphql::Request::builder()
                    .query(r#"{allProducts{variation {id}id}}"#)
                    .build(),
            )
            .collect::<Vec<_>>()
            .await;

        assert_eq!(collect[0], response);
        mock.assert();
        Ok(())
    }
}
