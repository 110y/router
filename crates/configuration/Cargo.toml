[package]
name = "configuration"
version = "0.1.0-prealpha.1"
authors = ["Apollo Graph, Inc. <opensource@apollographql.com>"]
edition = "2018"
license-file = "./LICENSE"

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
serde_yaml = "0.8.19"
serde = { version = "1.0.130", features = ["derive"] }
thiserror = "1.0.28"
maplit = "1.0.2"
typed-builder = "0.9.0"

[dev-dependencies]

