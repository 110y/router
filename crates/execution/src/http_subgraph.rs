use std::pin::Pin;

use bytes::Bytes;
use futures::prelude::*;

use crate::{
    FetchError, GraphQLFetcher, GraphQLPatchResponse, GraphQLPrimaryResponse, GraphQLRequest,
    GraphQLResponse, GraphQLResponseStream,
};

type BytesStream =
    Pin<Box<dyn futures::Stream<Item = Result<bytes::Bytes, FetchError>> + std::marker::Send>>;

/// A fetcher for subgraph data that uses http.
/// Streaming via chunking is supported.
#[derive(Debug)]
pub struct HttpSubgraphFetcher {
    service: String,
    url: String,
    http_client: reqwest::Client,
}

impl HttpSubgraphFetcher {
    /// Construct a new http subgraph fetcher that will fetch from the supplied URL.
    pub fn new(service: String, url: String) -> Self {
        HttpSubgraphFetcher {
            service,
            url,
            http_client: reqwest::Client::new(),
        }
    }

    fn request_stream(&self, request: GraphQLRequest) -> BytesStream {
        // Perform the actual request and start streaming.
        // Reqwest doesn't care if there is only one response, in this case it'll be a stream of one element.
        let service = self.service.to_owned();
        self.http_client
            .post(self.url.clone())
            .json(&request)
            .send()
            // We have a future for the response, convert it to a future of the stream.
            .map_ok(|r| r.bytes_stream().boxed())
            // Convert the entire future to a stream, at this point we have a stream of a result of a single stream
            .into_stream()
            // Flatten the stream
            .flat_map(|result| match result {
                Ok(s) => s,
                Err(err) => stream::iter(vec![Err(err)]).boxed(),
            })
            .map_err(move |err| FetchError::ServiceError {
                service: service.to_owned(),
                reason: err.to_string(),
            })
            .boxed()
    }

    fn map_to_graphql(&self, bytes_stream: BytesStream) -> GraphQLResponseStream {
        // Map the stream of bytes to our response type.
        let service = self.service.to_owned();
        let response_stream = bytes_stream
            // We want to know if this is the primary response or not, so attach the response count
            .enumerate()
            .map(move |(index, result)| {
                let service = service.to_owned();
                match result {
                    Ok(bytes) if { index == 0 } => to_primary(service, &bytes),
                    Ok(bytes) => to_patch(service, &bytes),
                    Err(err) => Err(err),
                }
            });
        response_stream.boxed()
    }
}

fn to_patch(service: String, bytes: &Bytes) -> Result<GraphQLResponse, FetchError> {
    serde_json::from_slice::<GraphQLPatchResponse>(&bytes)
        .map(GraphQLResponse::Patch)
        .map_err(move |err| FetchError::ServiceError {
            service,
            reason: err.to_string(),
        })
}

fn to_primary(service: String, bytes: &Bytes) -> Result<GraphQLResponse, FetchError> {
    serde_json::from_slice::<GraphQLPrimaryResponse>(&bytes)
        .map(GraphQLResponse::Primary)
        .map_err(move |err| FetchError::ServiceError {
            service,
            reason: err.to_string(),
        })
}

impl GraphQLFetcher for HttpSubgraphFetcher {
    /// Using reqwest fetch a stream of graphql results.
    fn stream(&self, request: GraphQLRequest) -> GraphQLResponseStream {
        let bytes_stream = self.request_stream(request);
        self.map_to_graphql(bytes_stream)
    }
}

#[cfg(test)]
mod tests {
    use httpmock::Method::POST;
    use httpmock::{MockServer, Regex};
    use serde_json::json;

    use super::*;

    #[tokio::test]
    async fn test_non_chunked() -> Result<(), Box<dyn std::error::Error>> {
        let response = GraphQLPrimaryResponse {
            data: json!({
              "allProducts": [
                {
                  "variation": {
                    "id": "OSS"
                  },
                  "id": "apollo-federation"
                },
                {
                  "variation": {
                    "id": "platform"
                  },
                  "id": "apollo-studio"
                }
              ]
            })
            .as_object()
            .unwrap()
            .to_owned(),
            has_next: Default::default(),
            errors: Default::default(),
            extensions: Default::default(),
        };

        let server = MockServer::start();
        let mock = server.mock(|when, then| {
            when.method(POST)
                .path("/graphql")
                .body_matches(Regex::new(".*").unwrap());
            then.status(200)
                .header("Content-Type", "application/json")
                .json_body_obj(&response);
        });
        let fetcher = HttpSubgraphFetcher::new("products".into(), server.url("/graphql"));
        let collect = fetcher
            .stream(
                GraphQLRequest::builder()
                    .query(r#"{allProducts{variation {id}id}}"#)
                    .build(),
            )
            .collect::<Vec<Result<GraphQLResponse, FetchError>>>()
            .await;

        assert_eq!(
            collect[0].as_ref().unwrap(),
            &GraphQLResponse::Primary(response)
        );
        mock.assert();
        Ok(())
    }
}
