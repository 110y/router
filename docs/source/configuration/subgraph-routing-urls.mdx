---
title: Configuring runtime subgraph routing URLs.
sidebar_title: Subgraph routing URLs
description: Overriding the default subgraph URLs used during query plan execution.
---

import { Link } from 'gatsby';

TODO

- By default, you don't need this.  We use the subgraph routing URLs that exist in the supergraph.
- Setting these overrides those values.

---

By default, the routing URLs for subgraphs are included in the supergraph during composition, either in Apollo Studio or in Rover.  In the event that you need to override the URL that is provided - for example, to account for changing network topography - you can change override this.

### Example

In this example the `accounts` subgraph URL is overridden to point to `http://localhost:2525` and the value which was specified during composition in either Apollo Studio or Rover will be ignored.

```yaml{3}
override_subgraph_url:
  # subgraph_name: new_url
  accounts: "http://localhost:2525"
```