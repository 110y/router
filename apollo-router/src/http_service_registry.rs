use crate::configuration::Configuration;
use crate::http_subgraph::HttpSubgraphFetcher;
use apollo_router_core::prelude::*;
use futures::Future;
use std::collections::HashMap;
use std::fmt;
use std::pin::Pin;
use std::sync::Arc;
use tower::Service;

/// Service registry that uses http to connect to subgraphs.
pub struct HttpServiceRegistry {
    services: HashMap<String, Arc<dyn NewSubgraphService>>,
}

impl fmt::Debug for HttpServiceRegistry {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        let mut debug = f.debug_tuple("HttpServiceRegistry");
        for name in self.services.keys() {
            debug.field(name);
        }
        debug.finish()
    }
}

impl HttpServiceRegistry {
    /// Create a new http service registry from a configuration.
    pub fn new(configuration: &Configuration) -> Self {
        Self {
            services: configuration
                .subgraphs
                .iter()
                .map(|(name, subgraph)| {
                    let fetcher =
                        HttpSubgraphFetcher::new(name.to_owned(), subgraph.routing_url.to_owned());
                    let cloner: Arc<dyn NewSubgraphService> = Arc::new(move || fetcher.clone());
                    (name.to_string(), cloner)
                })
                .collect(),
        }
    }

    // test dynamically adding a Service
    pub fn add_service<S>(&mut self, name: String, service: S)
    where
        S: Clone + SubgraphService,
    {
        let cloner: Arc<dyn NewSubgraphService> = Arc::new(move || service.clone());
        self.services.insert(name, cloner);
    }
}

impl graphql::ServiceRegistry for HttpServiceRegistry {
    fn get(&self, service: &str) -> Option<Box<dyn SubgraphService>> {
        self.services.get(service).map(|a| a.new_service())
    }

    fn has(&self, service: &str) -> bool {
        self.services.get(service).is_some()
    }
}

impl Service<(String, graphql::Request)> for HttpServiceRegistry {
    type Response = graphql::Response;

    type Error = graphql::FetchError;

    type Future = Pin<Box<dyn Future<Output = Result<Self::Response, Self::Error>>>>;

    fn poll_ready(
        &mut self,
        _cx: &mut std::task::Context<'_>,
    ) -> std::task::Poll<Result<(), Self::Error>> {
        std::task::Poll::Ready(Ok(()))
    }

    fn call(&mut self, (service_name, request): (String, graphql::Request)) -> Self::Future {
        let mut service = self
            .services
            .get(&service_name)
            .map(|a| &**a)
            .unwrap()
            .new_service();

        service.call(request)
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::configuration::Configuration;

    #[test]
    fn test_from_string() {
        let config =
            serde_yaml::from_str::<Configuration>(include_str!("testdata/supergraph_config.yaml"))
                .unwrap();
        let registry = HttpServiceRegistry::new(&config);
        assert!(registry.get("products").is_some())
    }
}
