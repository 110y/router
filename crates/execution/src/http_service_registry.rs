use std::collections::HashMap;

use configuration::Configuration;

use crate::http_subgraph::HttpSubgraphFetcher;
use crate::{GraphQLFetcher, ServiceRegistry};

/// Service registry that uses http to connect to subgraphs.
#[derive(Debug)]
pub struct HttpServiceRegistry {
    services: HashMap<String, Box<dyn GraphQLFetcher>>,
}

impl HttpServiceRegistry {
    /// Create a new http service registry from a configuration.
    pub fn new(configuration: &Configuration) -> Self {
        Self {
            services: configuration
                .subgraphs
                .iter()
                .map(|(name, subgraph)| {
                    let fetcher: Box<dyn GraphQLFetcher> = Box::new(HttpSubgraphFetcher::new(
                        name.to_owned(),
                        subgraph.routing_url.to_owned(),
                    ));
                    (name.to_string(), fetcher)
                })
                .collect(),
        }
    }
}

impl ServiceRegistry for HttpServiceRegistry {
    fn get(&self, service: &str) -> Option<&(dyn GraphQLFetcher)> {
        self.services.get(service).map(|a| &**a)
    }

    fn has(&self, service: &str) -> bool {
        self.services.get(service).is_some()
    }
}

#[cfg(test)]
mod tests {
    use configuration::Configuration;

    use crate::http_service_registry::HttpServiceRegistry;
    use crate::ServiceRegistry;

    #[test]
    fn test_from_string() {
        let config =
            serde_yaml::from_str::<Configuration>(include_str!("testdata/supergraph_config.yaml"))
                .unwrap();
        let registry = HttpServiceRegistry::new(&config);
        assert!(registry.get("products").is_some())
    }
}
