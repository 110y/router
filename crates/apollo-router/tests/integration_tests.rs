use apollo_router::configuration::Configuration;
use apollo_router::http_service_registry::HttpServiceRegistry;
use apollo_router::http_subgraph::HttpSubgraphFetcher;
use apollo_router_core::prelude::*;
use futures::prelude::*;
use maplit::hashmap;
use serde_json::to_string_pretty;
use std::collections::hash_map::Entry;
use std::collections::HashMap;
use std::sync::Arc;

#[ctor::ctor]
fn init() {
    let _ = env_logger::builder().is_test(true).try_init();
}

macro_rules! assert_federated_response {
    ($query:expr, $service_requests:expr $(,)?) => {
        init();
        let request = graphql::Request::builder()
            .query($query)
            .variables(Arc::new(
                vec![
                    ("topProductsFirst".to_string(), 2.into()),
                    ("reviewsForAuthorAuthorId".to_string(), 1.into()),
                ]
                .into_iter()
                .collect(),
            ))
            .build();
        let (mut actual, registry) = query_rust(request.clone());
        let mut expected = query_node(request.clone());

        let actual = actual.next().await.unwrap();
        let expected = expected.next().await.unwrap();
        eprintln!("actual: {}", to_string_pretty(&actual).unwrap());
        eprintln!("expected: {}", to_string_pretty(&expected).unwrap());

        // The current implementation does not cull extra properties that should not make is to the
        // output yet, so we check that the nodejs implementation returns a subset of the
        // output of the rust output.
        assert!(expected.data.is_subset(&actual.data));
        assert_eq!(registry.totals(), $service_requests);
    };
}

#[tokio::test]
async fn basic_request() {
    assert_federated_response!(
        r#"{ topProducts { name } }"#,
        hashmap! {
            "products".to_string()=>1,
        },
    );
}

#[tokio::test]
async fn basic_composition() {
    assert_federated_response!(
        r#"{ topProducts { upc name reviews {id product { name } author { id name } } } }"#,
        hashmap! {
            "products".to_string()=>2,
            "reviews".to_string()=>1,
            "accounts".to_string()=>1,
        },
    );
}

#[tokio::test]
async fn basic_mutation() {
    assert_federated_response!(
        r#"mutation {
              createProduct(upc:"8", name:"Bob") {
                upc
                name
                reviews {
                  body
                }
              }
              createReview(upc: "8", id:"100", body: "Bif"){
                id
                body
              }
            }"#,
        hashmap! {
            "products".to_string()=>1,
            "reviews".to_string()=>2,
        },
    );
}

#[tokio::test]
async fn variables() {
    assert_federated_response!(
        r#"
            query ExampleQuery($topProductsFirst: Int, $reviewsForAuthorAuthorId: ID!) {
                topProducts(first: $topProductsFirst) {
                    name
                    reviewsForAuthor(authorID: $reviewsForAuthorAuthorId) {
                        body
                        author {
                            id
                            name
                        }
                    }
                }
            }
            "#,
        hashmap! {
            "products".to_string()=>1,
            "reviews".to_string()=>1,
            "accounts".to_string()=>1,
        },
    );
}

#[tokio::test]
async fn missing_variables() {
    let request = graphql::Request::builder()
        .query(
            r#"
                query ExampleQuery($missingVariable: Int, $yetAnotherMissingVariable: ID!) {
                    topProducts(first: $missingVariable) {
                        name
                        reviewsForAuthor(authorID: $yetAnotherMissingVariable) {
                            body
                        }
                    }
                }
                "#,
        )
        .build();
    let (response, _) = query_rust(request.clone());
    let data = response
        .flat_map(|x| stream::iter(x.errors))
        .collect::<Vec<_>>()
        .await;
    let expected = vec![
        graphql::FetchError::ValidationMissingVariable {
            name: "yetAnotherMissingVariable".to_string(),
        }
        .to_graphql_error(None),
        graphql::FetchError::ValidationMissingVariable {
            name: "missingVariable".to_string(),
        }
        .to_graphql_error(None),
    ];
    assert!(data.iter().all(|x| expected.contains(x)));
}

fn query_node(request: graphql::Request) -> graphql::ResponseStream {
    let nodejs_impl =
        HttpSubgraphFetcher::new("federated".into(), "http://localhost:4000/graphql".into());
    nodejs_impl.stream(request)
}

fn query_rust(
    request: graphql::Request,
) -> (graphql::ResponseStream, Arc<CountingServiceRegistry>) {
    let schema: graphql::Schema = include_str!("fixtures/supergraph.graphql").parse().unwrap();
    let planner = graphql::HarmonizerQueryPlanner::new(&schema);
    let config =
        serde_yaml::from_str::<Configuration>(include_str!("fixtures/supergraph_config.yaml"))
            .unwrap();
    let registry = Arc::new(CountingServiceRegistry::new(HttpServiceRegistry::new(
        &config,
    )));
    let federated =
        graphql::FederatedGraph::new(Box::new(planner), registry.clone(), Arc::new(schema));
    (federated.stream(request), registry)
}

#[derive(Debug)]
struct CountingServiceRegistry {
    counts: Arc<parking_lot::Mutex<HashMap<String, usize>>>,
    delegate: HttpServiceRegistry,
}

impl CountingServiceRegistry {
    fn new(delegate: HttpServiceRegistry) -> CountingServiceRegistry {
        CountingServiceRegistry {
            counts: Arc::new(parking_lot::Mutex::new(HashMap::new())),
            delegate,
        }
    }

    fn totals(&self) -> HashMap<String, usize> {
        self.counts.lock().clone()
    }
}

impl ServiceRegistry for CountingServiceRegistry {
    fn get(&self, service: &str) -> Option<&dyn graphql::Fetcher> {
        let mut counts = self.counts.lock();
        match counts.entry(service.to_owned()) {
            Entry::Occupied(mut e) => {
                *e.get_mut() += 1;
            }
            Entry::Vacant(e) => {
                e.insert(1);
            }
        }
        self.delegate.get(service)
    }

    fn has(&self, service: &str) -> bool {
        self.delegate.has(service)
    }
}
