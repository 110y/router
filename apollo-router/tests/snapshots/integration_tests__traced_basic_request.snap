---
source: apollo-router/tests/integration_tests.rs
assertion_line: 94
expression: get_spans()

---
{
  "name": "integration_tests::root",
  "record": {
    "entries": [],
    "metadata": {
      "name": "root",
      "target": "integration_tests",
      "level": "INFO",
      "module_path": "integration_tests",
      "fields": {
        "names": []
      }
    }
  },
  "children": {
    "apollo_router_core::naive_introspection::naive_introspection_population": {
      "name": "apollo_router_core::naive_introspection::naive_introspection_population",
      "record": {
        "entries": [],
        "metadata": {
          "name": "naive_introspection_population",
          "target": "apollo_router_core::naive_introspection",
          "level": "INFO",
          "module_path": "apollo_router_core::naive_introspection",
          "fields": {
            "names": []
          }
        }
      },
      "children": {}
    },
    "apollo_router_core::services::router_service::query_parsing": {
      "name": "apollo_router_core::services::router_service::query_parsing",
      "record": {
        "entries": [],
        "metadata": {
          "name": "query_parsing",
          "target": "apollo_router_core::services::router_service",
          "level": "INFO",
          "module_path": "apollo_router_core::services::router_service",
          "fields": {
            "names": []
          }
        }
      },
      "children": {}
    },
    "apollo_router_core::query_planner::router_bridge_query_planner::plan": {
      "name": "apollo_router_core::query_planner::router_bridge_query_planner::plan",
      "record": {
        "entries": [],
        "metadata": {
          "name": "plan",
          "target": "apollo_router_core::query_planner::router_bridge_query_planner",
          "level": "DEBUG",
          "module_path": "apollo_router_core::query_planner::router_bridge_query_planner",
          "fields": {
            "names": []
          }
        }
      },
      "children": {}
    },
    "apollo_router_core::services::execution_service::execute": {
      "name": "apollo_router_core::services::execution_service::execute",
      "record": {
        "entries": [],
        "metadata": {
          "name": "execute",
          "target": "apollo_router_core::services::execution_service",
          "level": "DEBUG",
          "module_path": "apollo_router_core::services::execution_service",
          "fields": {
            "names": []
          }
        }
      },
      "children": {
        "apollo_router_core::services::execution_service::execution": {
          "name": "apollo_router_core::services::execution_service::execution",
          "record": {
            "entries": [],
            "metadata": {
              "name": "execution",
              "target": "apollo_router_core::services::execution_service",
              "level": "INFO",
              "module_path": "apollo_router_core::services::execution_service",
              "fields": {
                "names": []
              }
            }
          },
          "children": {
            "apollo_router_core::query_planner::fetch": {
              "name": "apollo_router_core::query_planner::fetch",
              "record": {
                "entries": [],
                "metadata": {
                  "name": "fetch",
                  "target": "apollo_router_core::query_planner",
                  "level": "INFO",
                  "module_path": "apollo_router_core::query_planner",
                  "fields": {
                    "names": []
                  }
                }
              },
              "children": {
                "apollo_router_core::query_planner::fetch::subfetch_stream": {
                  "name": "apollo_router_core::query_planner::fetch::subfetch_stream",
                  "record": {
                    "entries": [
                      [
                        "service",
                        "products"
                      ]
                    ],
                    "metadata": {
                      "name": "subfetch_stream",
                      "target": "apollo_router_core::query_planner::fetch",
                      "level": "INFO",
                      "module_path": "apollo_router_core::query_planner::fetch",
                      "fields": {
                        "names": [
                          "service"
                        ]
                      }
                    }
                  },
                  "children": {
                    "apollo_router_core::query_planner::fetch::make_variables": {
                      "name": "apollo_router_core::query_planner::fetch::make_variables",
                      "record": {
                        "entries": [],
                        "metadata": {
                          "name": "make_variables",
                          "target": "apollo_router_core::query_planner::fetch",
                          "level": "DEBUG",
                          "module_path": "apollo_router_core::query_planner::fetch",
                          "fields": {
                            "names": []
                          }
                        }
                      },
                      "children": {}
                    },
                    "apollo_router_core::query_planner::fetch::subfetch_stream": {
                      "name": "apollo_router_core::query_planner::fetch::subfetch_stream",
                      "record": {
                        "entries": [],
                        "metadata": {
                          "name": "subfetch_stream",
                          "target": "apollo_router_core::query_planner::fetch",
                          "level": "INFO",
                          "module_path": "apollo_router_core::query_planner::fetch",
                          "fields": {
                            "names": []
                          }
                        }
                      },
                      "children": {
                        "reqwest_tracing::middleware::HTTP request": {
                          "name": "reqwest_tracing::middleware::HTTP request",
                          "record": {
                            "entries": [
                              [
                                "http.method",
                                "POST"
                              ],
                              [
                                "http.scheme",
                                "http"
                              ],
                              [
                                "http.host",
                                "localhost"
                              ],
                              [
                                "net.host.port",
                                "4003"
                              ],
                              [
                                "otel.kind",
                                "client"
                              ],
                              [
                                "otel.name",
                                "POST /graphql"
                              ],
                              [
                                "otel.status_code",
                                "OK"
                              ],
                              [
                                "http.status_code",
                                200
                              ],
                              [
                                "http.user_agent",
                                ""
                              ]
                            ],
                            "metadata": {
                              "name": "HTTP request",
                              "target": "reqwest_tracing::middleware",
                              "level": "INFO",
                              "module_path": "reqwest_tracing::middleware",
                              "fields": {
                                "names": [
                                  "http.method",
                                  "http.scheme",
                                  "http.host",
                                  "net.host.port",
                                  "otel.kind",
                                  "otel.name",
                                  "otel.status_code",
                                  "http.user_agent",
                                  "http.status_code",
                                  "error.message",
                                  "error.cause_chain"
                                ]
                              }
                            }
                          },
                          "children": {}
                        },
                        "apollo_router::reqwest_subgraph_service::aggregate_response_data": {
                          "name": "apollo_router::reqwest_subgraph_service::aggregate_response_data",
                          "record": {
                            "entries": [],
                            "metadata": {
                              "name": "aggregate_response_data",
                              "target": "apollo_router::reqwest_subgraph_service",
                              "level": "DEBUG",
                              "module_path": "apollo_router::reqwest_subgraph_service",
                              "fields": {
                                "names": []
                              }
                            }
                          },
                          "children": {}
                        },
                        "apollo_router::reqwest_subgraph_service::parse_subgraph_response": {
                          "name": "apollo_router::reqwest_subgraph_service::parse_subgraph_response",
                          "record": {
                            "entries": [],
                            "metadata": {
                              "name": "parse_subgraph_response",
                              "target": "apollo_router::reqwest_subgraph_service",
                              "level": "DEBUG",
                              "module_path": "apollo_router::reqwest_subgraph_service",
                              "fields": {
                                "names": []
                              }
                            }
                          },
                          "children": {}
                        }
                      }
                    },
                    "apollo_router_core::query_planner::fetch::response_insert": {
                      "name": "apollo_router_core::query_planner::fetch::response_insert",
                      "record": {
                        "entries": [],
                        "metadata": {
                          "name": "response_insert",
                          "target": "apollo_router_core::query_planner::fetch",
                          "level": "DEBUG",
                          "module_path": "apollo_router_core::query_planner::fetch",
                          "fields": {
                            "names": []
                          }
                        }
                      },
                      "children": {}
                    }
                  }
                }
              }
            }
          }
        }
      }
    },
    "apollo_router_core::services::router_service::format_response": {
      "name": "apollo_router_core::services::router_service::format_response",
      "record": {
        "entries": [],
        "metadata": {
          "name": "format_response",
          "target": "apollo_router_core::services::router_service",
          "level": "DEBUG",
          "module_path": "apollo_router_core::services::router_service",
          "fields": {
            "names": []
          }
        }
      },
      "children": {}
    }
  }
}
